Disable automatic acknowledgement

akarl opened the pull request:
Adds functionality for disabling the automatic acknowledgement after handler execution.

This enables a user to have more fine-grained control over the acknowledgements.

bombsimon commented on bindings_test.go:
```diff
@@ -109,7 +109,7 @@ func TestTopic(t *testing.T) {
 					}
 					assert.Equal(t, tc.request, body, "correct request body")
-				case <-time.After(10 * time.Millisecond):
+				case <-time.After(10 * time.Second):
```
Isn't this a bit much, feels like `1 * time.Second` would be enough?

akarl replied:
It probably is, unless the Travis instance is particularly slow.

bombsimon commented:
Actually, I have a question.

There's already a concept on the `ConsumeSettings`:
https://github.com/bombsimon/amqp-rpc/blob/f512513c30a57b02e9297f07a8ccf68ff2721161/connection.go#L51

When using the default default settings for the server, this is always false:
https://github.com/bombsimon/amqp-rpc/blob/f512513c30a57b02e9297f07a8ccf68ff2721161/server.go#L116-L122

But for the client (which might be more reasonble) this is always true:
https://github.com/bombsimon/amqp-rpc/blob/f512513c30a57b02e9297f07a8ccf68ff2721161/client.go#L246-L253

The `autoAck` you implemented only does anything if the setting on the `ConsumeSettings` didn't take action (and was suppose to auto ack). The reason for this is that if a panic occurs (and is recovered by the middleware) the `amqp.Channel` will not ack (if it's set to auto ack).

So now I have a few questions. Is it really a good idea to introduce a new/different/yet another ack concept? What about just improve the settings interface (which we talked about) or just use the `WithConsumeSettings` available on both the client and the server and make it homogeneous?

```go
s := NewServer(url).
    WithConsumeSettings({
        AutoAck: false,
    })

// ...

if !aac.IsHandled() && !s.consumeSettings.AutoAck {
    // ack for the user
}
```

Also I'm wondering, what happens to the deliveries on the server today since we don't have the auto ack in the same way we do on the client? Is this a bug?

bombsimon commented:
@akarl Yeah I think I understand everything now and when I read the code again I see that we set `AutoAck` to false on the server but then we do this unconditional `d.Ack()` with the `AckAwareAcknowledger` (until now with your fix). I thought it was something for panics but that's in the middleware:
https://github.com/bombsimon/amqp-rpc/blob/f512513c30a57b02e9297f07a8ccf68ff2721161/middleware/panic_recovery.go#L43-L45

Isn't this a bit weird behaviour? Sure, we achieve the "At Least Once Delivery" behaviour since we ensure to only ack **after** the handler, but we do it a bit greedy in my opinion. I think it would be a better idea to do nothing if auto ack is false (like today) and let RabbitMQ handle if it's true. We could then add a generic middleware handling the ack. The only problem is that the default amqp acknowledger is unaware of if it's been handled and that's why I made `AckAwareAcknowledger`. I'm not sure how we would expose the `IsHandled()` publicly unless we type cast the `delivery.Acknowledger` in the middleware or create a new interface. We need to change something to avoid circular dependencies either way.

You think this is a god idea or do you still want to use a separate condition/struct field for this?

akarl commented:
This was my first thought, but that would break some backwards compatibility. Adding an included middleware for automatic acknowledgements would be an option. I'll see what I can come up with.

bombsimon commented:
Sounds great but don’t over do it if you think this is sufficient!

Since the server always adds the `AckAwareAcknowkedger` to the delivery an idea would be to just do the `IsHandled` check in’s each ack/nack/reject function?

akarl commented:
> Since the server always adds the AckAwareAcknowkedger to the delivery an idea would be to just do the IsHandled check in’s each ack/nack/reject function?

We could do that, but I think that's a bit too magic. If the user want's that error, what then?

bombsimon commented:
Yeah, true! I don’t even know what happens if you try to ack an acked message. :)

Just to keep writing (bad) ideas, another idea would be to add a field like `AckAfterHandle` and use that one just like you did but avoid introducing new (private) fields on the server. 

I’m down for whatever but like you said, this should be up to the user to decide and handle!

akarl commented:
> Yeah, true! I don’t even know what happens if you try to ack an acked message. :)

You'll get an error with something like "unknown delivery tag"

> Just to keep writing (bad) ideas, another idea would be to add a field like AckAfterHandle and use that one just like you did but avoid introducing new (private) fields on the server.

Yea that's almost the same thing as in the previous commit.

akarl commented:
Check the updated diff, and tell me what you think. It breaks (a bit) of backwards compatibility because we go from at least once to at most once delivery (per default). Changing that would mean that the user should disable `AutoAck` and add a middleware or do it manually in their handlers.

bombsimon commented on server.go:
```diff
@@ -407,25 +417,9 @@ func (s *Server) runHandler(handler HandlerFunc, deliveries <-chan amqp.Delivery
-		// Use the default provided Acknowledger for the delivery
-		// (amqp.Channel) and add our ack aware acknowledger which can tell if
-		// a message has been acknowledged (ack, nack or rejected).
-		aac := ackAwareChannel{
```
This is the only place where `ackAwareChannel` is used so if you want to remove it I guess we should remote the [whole file](https://github.com/bombsimon/amqp-rpc/blob/master/acknowledger.go) (and it's tests) as well. But should we? I think it's nice and if it's used you can type cast to an interface and check handled. If you add both our bundled middlewares now (panic recovery which will nack and auto ack which will ack) you'll get an error from the ack middleware. It's no biggie but it feels like bad design to ship with two middlewares that can't even co-exist.

akarl replied:
> If you add both our bundled middlewares now (panic recovery which will nack and auto ack which will ack) you'll get an error from the ack middleware.

Ahh, yes. But the `PanicRecovery` middleware won't work together with `AutoAck` anyway so I'll just remove the `Reject` and simplify it.

bombsimon replied:
👍 I guess this is why the `HandlerCrashedHeader` constant is there, so that users can look at the header and do ack/nack accordingly.

bombsimon commented on bindings_test.go:
```diff
@@ -56,72 +53,37 @@ func TestFanout(t *testing.T) {
 	stop := startAndWait(s)
 	defer stop()
 
-	cases := []struct {
```
Are all these cases still covered? Shouldn't this be a separate commit for something else (other than ack)?

akarl replied:
Yes! There where some problems with the way this test was implemented and it was causing slowdowns. We don't really need to test AMQP, we just need to make sure we did the binding correctly. This proves that because only a topic exchange behaves this way.

But I could put it in another PR if you want.

bombsimon replied:
Ok, cool! Nah keep it in this PR! Was just the commit (message) I was nit-picking about, but no biggie. Maybe this resolves #66? Would be great!

akarl replied:
It should, but let's see.

bombsimon commented on testing.go:
```diff
@@ -1,13 +1,22 @@
 	"github.com/streadway/amqp"
 )
 
+const (
```
This is about time, nice!

bombsimon commented on testing.go:
```diff
@@ -62,3 +71,63 @@ func startAndWait(s *Server) func() {
+	_ = resp.Body.Close()
+}
+
+func closeAllConnections() {
```
Wow, nice to do this via the HTTP API! 🎉

bombsimon commented on testing.go:
```diff
@@ -1,13 +1,22 @@
+	"fmt"
 	"net"
+	"net/http"
+	url2 "net/url"
```
Haha, instead of changing the variable name? Or is this IntelliJ trying to be smart? :)

bombsimon commented on middleware/ack.go:
```diff
@@ -0,0 +1,22 @@
+package middleware
```
I think a middleware like this is a great "compromise" which make it easy to turn of auto-ack but still acknowledge every delivery like this! Linked to my comment about the `ackAwareAcknowledger`, shouldn't at least we as creators of this handle it?

```go
type knowsIsHandled interface {
    IsHandled() bool
}

// ...
if aaa, ok := d.Acknowledger.(knowsIsHandled); ok {
    if aaa.IsHandled() {
        return
    }
}

if err := d.Ack(false); err != nil {
    fmt.Printf("could not ack delivery %s: %v", d.CorrelationId, err)
}

// ...
```

bombsimon commented:
I've removed the total-coverage check. I don't think it's helpful to fail because -0.1% anyway.

akarl commented:
I've updated the `Acknowledger`, implemented the check for handled acks in the Ack middleware. Removed the `Reject` from `PanicRecovery` because it won't work when `AutoAck` is set.

bombsimon commented:
I think this is the final one now! :) I didn't think of exposing the `AckAwareAcknowledger` because of circular dependencies but I'm stupid, the middleware won't be used from the `amqp` package so that won't happen. With that said, this is a nice addition!

akarl commented:
Yea, sorry about that. Since the reason for this PR is to be able to give the user more control I wanted to make the middleware better as well.

bombsimon commented:
I re-added the diff check but removed it as required. This way we can see the change but you should still be able to merge.

bombsimon commented:
Oh noes, a new intermittent error in test. :(

```text
=== RUN   TestServerReconnect
closed 127.0.0.1:37364 -> 127.0.0.1:5672
closed 127.0.0.1:37366 -> 127.0.0.1:5672
closed 127.0.0.1:37368 -> 127.0.0.1:5672
closed 127.0.0.1:37370 -> 127.0.0.1:5672
2019/08/28 14:48:27 client: got error: Exception (320) Reason: "CONNECTION_FORCED - Closed via management plugin", will reconnect in 0.5 second(s)
2019/08/28 14:48:27 server: got error: Exception (320) Reason: "CONNECTION_FORCED - Closed via management plugin", will reconnect in 0.5 second(s)
--- FAIL: TestServerReconnect (10.09s)
    server_test.go:148: 
        	Error Trace:	server_test.go:148
        	Error:      	Received unexpected error:
        	            	request timed out
        	Test:       	TestServerReconnect
```

akarl commented:
The latest commit removed all coverage on the retries for the Client. I'll see if I can fix that.

akarl commented on client_test.go:
```diff
@@ -56,32 +58,57 @@ func TestClientConfig(t *testing.T) {
 
-	r := NewRequest().WithBody("client testing").WithResponse(false)
-	r.numRetries = client.maxRetries + 1
+	_, err = client.Send(NewRequest().WithResponse(false))
```
Heh
